Feature/agent ucc listener onboarding

morantr opened the pull request:
Any test case triggered by the UCC application starts with dev_reset_defaults which "tells" the agent that it should go back to its default state, meaning unconfigured and not onboarded.
Next, onboarding is triggered as part of the dev_set_config command.

This PR completes the agent wfa_ca listener (agent ucc listener) by adding agent implementation to dev_reset_default and dev_set_config.

The complete prplMesh agent flow starts with the UCC application triggering
dev_reset_default, which results in a full slave reset triggered by the backhaul manager as a response to the ucc agent's request.
The ucc agent now waits in BACKHAUL_SEND_ENABLE state until the dev_set_config command from the UCC application is received, thus blocking the agent from starting onboarding + autoconfig flows.

Next, the UCC sends "dev_set_config,name,IntelCTTDUT,program,map,backhaul,eth" command (can be either wired or wireless), for which the ucc agent listener replies with RUNNING, and sets the onboarding and onboarding to "in progress" state, continue with the onboarding with the requested radio/eth interface.

The son_slave now unblocks and proceeds with the state machine, which results in receiving M2 which unblocks the ucc listener the return "COMPLETE" status to ucc.

Fixes #336

arnout commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Fix dev_reset_default for agent](https://trello.com/c/Vcn26ZWp/12-fix-devresetdefault-for-agent)

morantr committed:
agent_ucc_listener: Add onboarding helper member variables

Preparative commit.
Add the following member variables to the agent_ucc_listener which will
be used on the onboarding process triggered by the UCC:

m_onboarding_state - Holds the onboarding process state.
It Will be set to 'true' upon receiving "DEV_RESET_DEFAULT" command
and change through the onboarding process which will end by a success
or a failure state.

m_selected_backhaul - Will hold the selected backhaul which has been
received on the "DEV_SET_CONFIG" command.

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

morantr committed:
agent_ucc_listener: Handle "DEV_RESET_DEFAULT"

Add initiation of onboarding process upon receiving "DEV_RESET_DEFAULT"
command from the UCC, by setting the onboarding state to
'eONBOARDING_RESET_TO_DEFAULT'.
Also, update the clear_configuration() fucnction description.

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

morantr committed:
Agent: Handle onboarding reset to default

Add get_and_update_onboarding_state() function to the agent_ucc_listener
that returns the onboarding state and increment the state in case the
state is eONBOARDING_RESET_TO_DEFAULT. This is since we want the
backhaul manager to reset only once since the query of that state which
has been also added to this commit, is on the backhaul work() function,
so if the state would be incremented, the backhaul will do infinite
resets.

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

morantr committed:
Agent: Stay in ENABLE state of the backhaul manager

After the Agent receives a DEV_RESET_DEFAULT command from the UCC, it is
expected to receive DEV_SET_CONFIG which will configure the interface
to bring up the link to the controller with.
Add code block that forces the backhaul manager to remain on ENABLE
state until the DEV_SET_CONFIG is received.

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

morantr committed:
agent_ucc_listener: Increment to eONBOARDING_IN_PROGRESS

Add getter function to selected backhaul which is being received on
DEV_SET_CONFIG command.
Also, increment agent_ucc_listener onboarding state if the selected
backhaul was set.
The set of the selected backhaul will be added in one of the followups
commits.

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

morantr committed:
tlvf: Add RUID to ACTION_BACKHAUL_REGISTER_REQUEST message

Preparative commit.
On a DEV_SET_CONFIG command, the UCC can set the te backhaul to
a specific RUID. In that case, the backhaul manager is required to use
this radio.
Add a RUID parameter to ACTION_BACKHAUL_REGISTER_REQUEST message
so the backhaul manager could use it when selecting the backhaul radio.

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

morantr committed:
backhaul manager: Add RUID from slave registration to socket struct

Add RUID from son_slave registration to relevant socket struct on
backhaul manager.

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

morantr committed:
backhaul manager: Select requested interface for link establishment

If the onboarding state changed to eONBOARDING_IN_PROGRESS, read
the requested backhaul interface from the agent_ucc_listener and
select it for link establishment.

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

morantr committed:
tlvf: Add ACTION_BACKHAUL_ONBOARDING_FINISHED_NOTIFICATION message

Preparative commit.
According to CAPI specification, the Agent should return a "COMPLETE"
status only when onboarding is finished. The onboarding finish is
defined when the Agent receives from the controller an
Auto-Configuration M2 message.

Add to the TLVF ACTION_BACKHAUL_ONBOARDING_FINISHED_NOTIFICATION message
which the son_slave will send to the backhaul manager when it receives
M2 message.

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

morantr committed:
backhaul manager: Update onboarding status

Add to agent_ucc_listener set_onboarding_status() function that updates
the agent_ucc_listener if the onboarding has succeeded or failed.

On the son_slave, send onboarding finish notification when receiving
M2 message.

On the backhaul manager, add a call to this function when receiving
ACTION_BACKHAUL_ONBOARDING_FINISHED_NOTIFICATION message
with success flag set to 'true' and on RESTART state with success flag
set to 'false'.

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

morantr committed:
agent_ucc_listener: Handle DEV_SET_CONFIG

Add implementation to handle_dev_set_config() function on the
agent_ucc_listener which extract the backhaul type from the command,
and block the thread until the onboarding state changes to SUCCESS or
FAILURE, return COMPLETE or ERROR status accordingly, back to the
UCC and reset the onboarding state to the initial state.

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

morantr committed:
agent_ucc_listener: Remove "eOnboarding_" prefix from eOnboardingState

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>
